nifi-commons/nifi-utils/src/main/java/org/apache/nifi/stream/io/util/TextLineDemarcator.java

```diff
@@ -90,58 +70,73 @@ public OffsetInfo nextOffsetInfo() {
-                for (i = this.index; i < this.bufferLength; i++) {
+                int i;
+                for (i = this.index; i < this.availableBytesLength; i++) {
                     byteVal = this.buffer[i];
-                    lineLength++;
-                    int crlfLength = computeEol(byteVal, i + 1);
-                    if (crlfLength > 0) {
-                        i += crlfLength;
-                        if (crlfLength == 2) {
-                            lineLength++;
-                        }
-                        offsetInfo = new OffsetInfo(this.offset, lineLength, crlfLength);
+
+                    if (byteVal == LF) {
+                        delimiterSize = previousByteVal == CR ? 2 : 1;
+                    } else if (previousByteVal == CR) {
+                        delimiterSize = 1;
+                        i--;
+                    }
+                    previousByteVal = byteVal;
+                    if (delimiterSize > 0) {
+                        this.index = i + 1;
+                        int size = Math.max(1, this.index - this.mark);
+                        offsetInfo = new OffsetInfo(this.offset, size, delimiterSize);
+                        this.offset += size;
                         if (startsWith != null) {
-                            token = this.extractDataToken(lineLength);
+                            data = this.extractDataToken(size);
                         }
                         this.mark = this.index;
-                        break lineLoop;
+                        break nextTokenLoop;
                     }
                 }
                 this.index = i;
+            } else {
+                delimiterSize = previousByteVal == CR || previousByteVal == LF ? 1 : 0;
+                if (offsetInfo == null) {
+                    int size = this.index - this.mark;
+                    if (size > 0) {
+                        offsetInfo = new OffsetInfo(this.offset, size, delimiterSize);
+                        this.offset += size;
+                    }
+                }
+                if (startsWith != null) {
+                    data = this.extractDataToken(this.index - this.mark);
+                }
             }
         }
-        // EOF where last char(s) are not CRLF.
-        if (lineLength > 0 && offsetInfo == null) {
-            offsetInfo = new OffsetInfo(this.offset, lineLength, 0);
-            if (startsWith != null) {
-                token = this.extractDataToken(lineLength);
-            }
-        }
-        this.offset += lineLength;
 
-        // checks if the new line starts with 'startsWith' chars
-        if (startsWith != null) {
+        if (startsWith != null && data != null) {
             for (int i = 0; i < startsWith.length; i++) {
                 byte sB = startsWith[i];
-                if (token != null && sB != token[i]) {
-                    offsetInfo.setStartsWithMatch(0);
+                if (sB != data[i]) {
```

Comment:
There is a potential `ArrayIndexOutOfBoundsException` here. If `startsWith` is longer than the last token that is extracted and the token does match the first `data.length` bytes of the `startsWith` array, then it will throw an `ArrayIndexOutOfBoundsException`. The following unit test will expose that:


```
@Test
    public void testStartsWithLongerThanLastToken() throws IOException {
        final byte[] inputData = "This is going to be a spectacular test\nThis is".getBytes(StandardCharsets.UTF_8);
        final byte[] startsWith = "This is going to be".getBytes(StandardCharsets.UTF_8);

        try (final InputStream is = new ByteArrayInputStream(inputData);
            final TextLineDemarcator demarcator = new TextLineDemarcator(is)) {

            final OffsetInfo first = demarcator.nextOffsetInfo(startsWith);
            assertNotNull(first);
            assertEquals(0, first.getStartOffset());
            assertEquals(39, first.getLength());
            assertEquals(1, first.getCrlfLength());
            assertTrue(first.isStartsWithMatch());

            final OffsetInfo second = demarcator.nextOffsetInfo(startsWith);
            assertNotNull(second);
            assertEquals(39, second.getStartOffset());
            assertEquals(7, second.getLength());
            assertEquals(0, second.getCrlfLength());
            assertFalse(second.isStartsWithMatch());
        }
    }

```